src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Raw.cs

```diff
@@ -0,0 +1,179 @@
+        public void WriteRawValue(ReadOnlySpan<byte> utf8Json, bool skipInputValidation = false)
+        {
+            if (!_options.SkipValidation)
+            {
+                ValidateWritingValue();
+            }
+
+            WriteRawValueInternal(utf8Json, skipInputValidation);
+        }
+
+        private void TranscodeAndWriteRawValue(ReadOnlySpan<char> json, bool skipInputValidation)
+        {
+            byte[]? tempArray = null;
+
+            // For performance, avoid obtaining actual byte count unless memory usage is higher than the threshold.
+            Span<byte> utf8Json = json.Length <= (JsonConstants.ArrayPoolMaxSizeBeforeUsingNormalAlloc / JsonConstants.MaxExpansionFactorWhileTranscoding) ?
+                // Use a pooled alloc.
+                tempArray = ArrayPool<byte>.Shared.Rent(json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding) :
+                // Use a normal alloc since the pool would create a normal alloc anyway based on the threshold (per current implementation)
+                // and by using a normal alloc we can avoid the Clear().
+                new byte[JsonReaderHelper.GetUtf8ByteCount(json)];
+
+            try
+            {
+                int actualByteCount = JsonReaderHelper.GetUtf8FromText(json, utf8Json);
+                utf8Json = utf8Json.Slice(0, actualByteCount);
+                WriteRawValue(utf8Json, skipInputValidation);
+            }
+            finally
+            {
+                if (tempArray != null)
+                {
+                    utf8Json.Clear();
+                    ArrayPool<byte>.Shared.Return(tempArray);
+                }
+            }
+        }
+
+        private void WriteRawValueInternal(ReadOnlySpan<byte> utf8Json, bool skipInputValidation)
+        {
+            int len = utf8Json.Length;
+
+            if (len == 0)
+            {
+                ThrowHelper.ThrowArgumentException(SR.ExpectedJsonTokens);
+            }
+            else if (len > JsonConstants.MaxRawValueLength)
+            {
+                ThrowHelper.ThrowArgumentException_ValueTooLarge(len);
+            }
+
+            if (skipInputValidation)
+            {
+                // Treat all unvalidated raw JSON value writes as string. If the payload is valid, this approach does
+                // not affect structural validation since a string token is equivalent to a complete object, array,
+                // or other complete JSON tokens when considering structural validation on subsequent writer calls.
+                // If the payload is not valid, then we make no guarantees about the structural validation of the final payload.
+                _tokenType = JsonTokenType.String;
+            }
+            else
+            {
+                // Utilize reader validation.
+                Utf8JsonReader reader = new(utf8Json);
+                while (reader.Read())
+                {
+                    _tokenType = reader.TokenType;
```

Comment:
Rather than keep doing this in a loop unnecessarily, why not set the `_tokenType` field outside the loop after `reader.Read()` returns false? Does that give the wrong value?

Reply:
Right, thanks. It gives the right value.